test/ret/cached_file_test.exs

```diff
@@ -31,4 +31,141 @@ defmodule Ret.CachedFileTest do
   test "vaccuum shouldn't fail" do
     CachedFile.vacuum()
   end
+
+  test "CachedFiles are vacuumed based on access time" do
+    with {:ok, _} <- CachedFile.fetch("aaa", write_to_path("aaa")),
+         {:ok, _} <- CachedFile.fetch("bbb", write_to_path("bbb")),
+         {:ok, _} <- CachedFile.fetch("ccc", write_to_path("ccc")) do
+      aaa = cached_file("aaa")
+      bbb = cached_file("bbb")
+      ccc = cached_file("ccc")
+
+      now = Timex.now()
+      one_day_from_now = shift(%{now: now, shift_options: [days: 1]})
+      one_week_from_now = shift(%{now: now, shift_options: [weeks: 1]})
+      three_weeks_from_now = shift(%{now: now, shift_options: [weeks: 3]})
+
+      Storage.fetch(aaa)
+      Storage.fetch_at(bbb, one_day_from_now)
+      Storage.fetch_at(ccc, three_weeks_from_now)
+
+      CachedFile.vacuum(%{expiration: one_week_from_now})
+
+      # The CachedFile has been vacuumed
+      assert cached_file("aaa") === nil
+      assert cached_file("bbb") === nil
+      assert %CachedFile{cache_key: "ccc"} = cached_file("ccc")
+
+      # The underlying asset has been vacuumed
+      assert {:error, "File not found"} = Storage.fetch(aaa.file_uuid, aaa.file_key, Storage.cached_file_path())
+      assert {:error, "File not found"} = Storage.fetch(bbb.file_uuid, bbb.file_key, Storage.cached_file_path())
+      assert {:ok, _meta, _stream} = Storage.fetch(ccc)
+    end
+  end
+
+  test "CachedFiles are migrated from expiring storage to cached storage" do
+    cache_key = "abc"
+
+    %CachedFile{file_uuid: file_uuid, file_key: file_key} = cached_file = put_file_in_expiring_storage(cache_key, "abc")
+
+    # This line ensures the file is stored in the expiring_file_path
+    {:ok, _meta, _stream} = Storage.fetch(file_uuid, file_key, Storage.expiring_file_path())
+    # This line ensures the file is NOT stored in the cached_file_path
+    {:error, _} = Storage.fetch(file_uuid, file_key, Storage.cached_file_path())
+    # This line causes the file to be copied to the cached_file_path
+    {:ok, _meta, _stream} = Storage.fetch(cached_file)
+    # This line ensures the file has been copied to the cached_file_path
+    {:ok, _meta, _stream} = Storage.fetch(file_uuid, file_key, Storage.cached_file_path())
+  end
+
+  test "CachedFiles are migrated from expiring storage to cached storage only once" do
```

Comment:
Technically, this test does not ensure that the migration happens only once. However, it _does_ ensure that if 500 agents try to access the file simultaneously, they will all succeed. I used println debugging to make sure that the file was only migrated once (and that the lock was working as intended), but I don't know how to test this programmatically.

Reply:
I think that's fine. We could do more sophisticated things to ensure that the migration only happens once, but it's enough that the code is exercised and that it doesn't fail.

Is it expensive to do this 500 times? I would think having 10 or 50 agents should be sufficient. Would be nice if the test itself didn't block for more than a few seconds.

Reply:
Happy to change it to 10 or 50 but it doesn't block for long at all -- less than a second on my machine. Here's a snippet of output that I had while testing:

```
{"got lock: migrate_b6e139fc-f671-4785-9a89-25d93c7a73a5",
 ~U[2021-03-18 04:43:16.301347Z]}
{"got lock: migrate_b6e139fc-f671-4785-9a89-25d93c7a73a5",
 ~U[2021-03-18 04:43:16.302087Z]}
{"got lock: migrate_b6e139fc-f671-4785-9a89-25d93c7a73a5",
 ~U[2021-03-18 04:43:16.302887Z]}
{"got lock: migrate_b6e139fc-f671-4785-9a89-25d93c7a73a5",
 ~U[2021-03-18 04:43:16.303970Z]}
{"got lock: migrate_b6e139fc-f671-4785-9a89-25d93c7a73a5",
 ~U[2021-03-18 04:43:16.304901Z]}
{"got lock: migrate_b6e139fc-f671-4785-9a89-25d93c7a73a5",
 ~U[2021-03-18 04:43:16.305694Z]}
{"got lock: migrate_b6e139fc-f671-4785-9a89-25d93c7a73a5",
 ~U[2021-03-18 04:43:16.306698Z]}
{"got lock: migrate_b6e139fc-f671-4785-9a89-25d93c7a73a5",
 ~U[2021-03-18 04:43:16.307423Z]}
{"got lock: migrate_b6e139fc-f671-4785-9a89-25d93c7a73a5",
 ~U[2021-03-18 04:43:16.308170Z]}
{"got lock: migrate_b6e139fc-f671-4785-9a89-25d93c7a73a5",
 ~U[2021-03-18 04:43:16.308892Z]}
{"got lock: migrate_b6e139fc-f671-4785-9a89-25d93c7a73a5",
 ~U[2021-03-18 04:43:16.309627Z]}
{"got lock: migrate_b6e139fc-f671-4785-9a89-25d93c7a73a5",
 ~U[2021-03-18 04:43:16.310488Z]}
{"got lock: migrate_b6e139fc-f671-4785-9a89-25d93c7a73a5",
 ~U[2021-03-18 04:43:16.311215Z]}
{"got lock: migrate_b6e139fc-f671-4785-9a89-25d93c7a73a5",
 ~U[2021-03-18 04:43:16.311948Z]}
{"got lock: migrate_b6e139fc-f671-4785-9a89-25d93c7a73a5",
 ~U[2021-03-18 04:43:16.312665Z]}
{"got lock: migrate_b6e139fc-f671-4785-9a89-25d93c7a73a5",
 ~U[2021-03-18 04:43:16.313565Z]}
{"got lock: migrate_b6e139fc-f671-4785-9a89-25d93c7a73a5",
 ~U[2021-03-18 04:43:16.314357Z]}
{"got lock: migrate_b6e139fc-f671-4785-9a89-25d93c7a73a5",
 ~U[2021-03-18 04:43:16.315083Z]}
 ```

Reply:
Ah, okay, as long as the test runs quickly, that's fine.